orttraining/orttraining/python/checkpointing_utils.py

```diff
@@ -1,143 +1,124 @@
-        
+        self.world_size = int(self.checkpoint_files[0].split('ZeRO')[1].split('.')[2]) + 1
+        assert len(self.checkpoint_files) == self.world_size, f"Could not find {self.world_size} files"
         self.weight_shape_map = dict()
-    
-    def _is_sharded(self, name):
-        if '_view_' in name:
-            return True
-        return False
-
-    def _has_fp16_weights(self, state_dict):
-        for k in state_dict.keys():
-            if k.endswith('_fp16'):
-                return True
-        return False
-    
-    def _split_moment_name(self, name):
+        self.sharded_params = set()
+
+    def _split_name(self, name):
         name_split = name.split('_view_')
-        if(len(name_split)>1):
+        if(len(name_split) > 1):
             view_num = int(name_split[1])
         else:
             view_num = None
-        weight_name = name_split[0].split('Moment_')[1][2:]
-        moment_num = int(name_split[0].split('Moment_')[1][0])
-        return moment_num, weight_name, view_num
+        optimizer_key = ''
+        fp16_key = ''
+        if name_split[0].startswith('Moment_1'):
+            optimizer_key = 'Moment_1_'
+        elif name_split[0].startswith('Moment_2'):
+            optimizer_key = 'Moment_2_'
+        elif name_split[0].startswith('Update_Count'):
+            optimizer_key = 'Update_Count_'
+        elif name_split[0].endswith('_fp16'):
+            fp16_key = '_fp16'
+        param_name = name_split[0]
+        if optimizer_key != '':
+            param_name = param_name.split(optimizer_key)[1]
+        param_name = param_name.split('_fp16')[0]
+        return param_name, optimizer_key, view_num, fp16_key
 
     def _update_weight_statistics(self, name, value):
-        self.weight_shape_map[name] = value.size() #original shape of tensor
-
-    def _reshape_tensors(self, state_dict, fp16):
-        for k,v in state_dict.items():
-            if k.startswith('Moment_'):
-                _, weight_name, _ = self._split_moment_name(k)
-                set_size = self.weight_shape_map[weight_name]    
-                state_dict[k] = v.reshape(set_size)
-                state_dict[weight_name] = state_dict[weight_name].reshape(set_size)
-        return state_dict
-  
+        if name not in self.weight_shape_map:
+            self.weight_shape_map[name] = value.size()  # original shape of tensor
+
+    def _reshape_tensor(self, key):
+        value = self.aggregate_state_dict[key]
+        weight_name, _, _, _ = self._split_name(key)
+        set_size = self.weight_shape_map[weight_name]
+        self.aggregate_state_dict[key] = value.reshape(set_size)
+
+    def _aggregate(self, param_dict):
+        for k, v in param_dict.items():
+            weight_name, optimizer_key, view_num, fp16_key = self._split_name(k)
+            if view_num is not None:
+                # parameter is sharded
+                param_name = optimizer_key + weight_name + fp16_key
```

Comment:
The 'fp16_key' denotes the suffix added to the parameter to declare that it was mixedprecision. Changing the variable name to improve understanding.

Reply:
Leaving this here for future reference. An assumption cannot be safely made that if the keyname ends with ```'fp_16'```, it will not be sharded. That is to say that if there is a state key that ends with ```'fp_16'```, it can be sharded.